### ISSUE TITLE
[P0][fork] hydra_recv_a treats EAGAIN as EOF → KV restore truncated → corrupted pool → decode fail at 37K ctx

### ISSUE THREAD

**Issue by ddvnguyen** [ddvnguyen/hydra_vortex#713]
## Summary
On the 6-turn A/B, turn6 (37,118 tokens) failed deterministically (3/3 runs) with 503. Engine trace shows the pre-prefill KV restore (STATE_GET → state_seq_set_data) **loses the tail of the 571.5 MiB state stream** because `hydra_recv_a` treats a transient `EAGAIN` on the socket as a clean end-of-file.

## Repro (deterministic)
6-turn same-session chat via core-A :19000 with force_mode=solo → turn6 (37.1K tok) always 503 `prefill_engine_terminal_error`.

## Engine trace (nodeA.log, VM 192.168.122.21)
```
E state_seq_set_data: error loading state: unexpectedly reached end of buffer
W srv  hydra_recv_a: hydra rpc: recv failed on fd=38 (0/16 bytes): Resource temporarily unavailable   <- EAGAIN!
E init_batch: failed to prepare attention ubatches
W decode: failed to find a memory slot for batch of size 512
E srv  process_sing: hydra: PREFILL slot=0 llama_decode failed at batch 33792
```
- State size at failure: 571.5 MiB (n_past=31363)
- Smaller restores (292-534 MiB) show the same EAGAIN warnings but recover; at ~571 MiB the truncation lands mid-struct and the pool is silently corrupted (no error propagation — downstream just fails at decode).

## Root cause
`hydra_recv_a` retry loop does not distinguish EAGAIN (would-block, must retry) from EOF (peer closed). A partial read followed by EAGAIN is treated as stream end → `state_seq_set_data` gets a short buffer → error, but the corrupted pool isn't flagged — subsequent PREFILL on the same slot runs on garbage KV.

## Impact
- Every session crossing ~35K+ ctx with shared-prefix restore path → hard 503 failure
- Corrupted pool is not detected/reported → silent data corruption risk for anything reusing that slot
- Blocks the whole M-Perf prefix-reuse work (#712) — restore path must be trustworthy first

## Fix direction
1. In `hydra_recv_a` (fork, tools RPC server path): treat EAGAIN/EWOULDBLOCK as retryable (with bounded retries/timeout), not stream-end.
2. Propagate restore failure: state_seq_set_data short-buffer must fail the RPC (and mark slot pool dirty) rather than leaving a corrupt pool for the next request.
3. Add regression test: force partial stream (small SO_RCVBUF or delayed sender) and assert restore fails loudly, not silently.

## Acceptance criteria
- [ ] turn6 A/B completes without 503 (evidence committed)
- [ ] EAGAIN on restore socket retries up to timeout instead of truncating
- [ ] Short-buffer restore fails loudly and slot is flagged dirty/quarantined
- [ ] Regression test for partial-stream restore added

**Comment by ddvnguyen** [comment]
## Live verify on v9721 (9583c5b37) — P100 test rig, 6-turn A/B rerun

**Shipped fix behavior confirmed:**
- EAGAIN no longer kills streams instantly — engine now logs typed failures (`recv failed ... Connection timed out` = 30s budget expired; distinct EOF logging).
- Regression test 6/6; version stamps `9721 (9583c5b37)`; sm_60 artifact byte-size-identical to proven v9720 build recipe.
- Turns 1-5 completed (t5 175.4s vs 228.1s on v9720 — run variance).

**Acceptance NOT met: turn6 (37.1K) still 503s.** Refined root-cause chain from traces:

1. Post-decode STATE_GET save SUCCEEDS every turn (653 MB, `rpc_status=0`).
2. Next turn's pre-PREFILL STATE_PUT restore **arrives truncated** — `state_seq_set_data: unexpectedly reached end of buffer` ~13ms after the save completes. Happens on EVERY turn (2-6), not just turn6 — small contexts survive it, 37K doesn't.
3. **Fork gap:** on restore failure in the pre-PREFILL path (task -1 STATE_PUT), the engine does NOT quarantine the slot — PREFILL proceeds on the half-restored pool.
4. Corrupt pool → `failed to find a memory slot for batch of size 512` → `llama_decode failed at batch 33792` (deterministic boundary).

**Remaining work split:**
- [ ] (core/unknown — follow-up issue) restore-stream sender truncation: coordinator STATE_PUT sends less than state size; needs bisect between core sender framing and fork reader.
- [ ] (fork — this issue's residual scope) slot quarantine on failed pre-PREFILL restore: mirror the DECODE_APPLY cleanup (clear tokens/checkpoints/KV cells, `llama_memory_seq_rm`, report to coordinator) in the PREFILL pre-restore path; fail the RPC loudly instead of prefilling on a dirty pool.

**Comment by ddvnguyen** [comment]
## Status update — fork fixes ready for review/merge (2026-08-29)

Two stacked fixes on the fork, branch `fix/713-slot-quarantine`:

**1. EAGAIN/EWOULDBLOCK retry in state-stream recv paths** (`9583c5b37`) — shared helper `common/hydra-socket-retry.h` used by both `refill()` and `hydra_recv_all()`; drain-on-POLLHUP; EINTR on poll+recv; 30s per-call budget. Regression test tests the real function (6 cases incl. POLLHUP drain), ctest 1/1 pass, sm_60 build clean. t2-reviewed (round 1: REQUEST_CHANGES → all 4 findings addressed in revision).

**2. Slot quarantine on STATE_PUT zero-read** (`75a1ceea3`) — `n_read == 0` branch now does `slot->prompt_clear(false)` (full cleanup: KV cells + tokens + logits, mirroring DECODE_APPLY) + loud `SRV_ERR`, so PREFILL cannot run on a dirty KV pool. `HYDRA_STATUS_ERROR` preserved → coordinator retry. Build PASS (build-hydra-dev). Lead-verified diff + build.

**Fork PR:** https://github.com/ddvnguyen/llama.cpp/pull/106 (base `hydra-fork`)

**Remaining steps (owner-gated):**
- [ ] Merge fork PR #106 into `hydra-fork`
- [ ] Bump `src/llama-cpp` submodule in hydra_vortex + close this issue (`Closes #713` PR)
- [ ] Live-rig verify on P100 restore path (240–650 MB blobs)
